pkg/action/action.go

```diff
@@ -366,6 +366,7 @@ func (cfg *Configuration) recordRelease(r *release.Release) {
 func (cfg *Configuration) Init(getter genericclioptions.RESTClientGetter, namespace, helmDriver string, log DebugLog) error {
 	kc := kube.New(getter)
 	kc.Log = log
+	kc.Namespace = namespace
```

Comment:
This code is a bit confusing and not ideal. I understand the desire to add the namespace here.

Without the `kc.Namespace` addition, the namespace is calculated inside the `kube.Client` here:

https://github.com/helm/helm/blob/b4b629c3c2a0e122d9332f8c436ddf9cf2d06968/pkg/kube/client.go#L315-L323

The factory that's used is setup here:

https://github.com/helm/helm/blob/b4b629c3c2a0e122d9332f8c436ddf9cf2d06968/pkg/kube/client.go#L97-L100

And all of that is from the [`cli.EnvSettings`](https://github.com/helm/helm/blob/b4b629c3c2a0e122d9332f8c436ddf9cf2d06968/pkg/cli/environment.go#L48).

It's a hard code path to follow.

All of this is because there are two (at least) namespaces in play. First, there is the namespace where the Helm Secret/ConfigMap, etc is stored. Second, there is the namespace the resources are installed into. Those can technically be separate.

In the original code the `getter` should have the namespace set on it. When `kc` communicates with the Kubernetes API it will use the namespace figured out from `getter` if it's not set by `kc.Namespace`.

In Helm this is figured out from the `settings` object...

https://github.com/helm/helm/blob/b4b629c3c2a0e122d9332f8c436ddf9cf2d06968/cmd/helm/helm.go#L75-L77

`settings` has a `SetNamespace` on it. So you could so something like...

```go
settings.SetNamespace("foo")
if err := actionConfig.Init(settings.RESTClientGetter(), settings.Namespace(), helmDriver, debug); err != nil {
...
```

This would setup the KubeClient to use the namespace you want to set programatically.

None of the current code is idea and I have to reorient myself to it each time. But, I think it needs a refactor rather than this small change. This small change would need some good QA testing before we could approve it. I can think of some use cases where it would break some SDK users and that's my concern.